File: lib/tools/date_tools.dart
import 'package:intl/intl.dart';

class DateTools {
  /// Returns week number for a date (https://en.wikipedia.org/wiki/ISO_week_date#Calculation)
  int getWeekNumber(DateTime date) {
    int dayOfYear = int.parse(DateFormat("D").format(date));
    int weekNumber = ((dayOfYear - date.weekday + 10) / 7).floor();
    if (weekNumber < 1) {
      weekNumber = getNumberOfWeeks(date.year - 1);
    } else if (weekNumber > getNumberOfWeeks(date.year)) {
      weekNumber = 1;
    }
    return weekNumber;
  }

  /// Returns number of weeks in a given year (https://en.wikipedia.org/wiki/ISO_week_date#Weeks_per_year)
  int getNumberOfWeeks(int year) {
    final december28 = DateTime(year, 12, 28);
    int dayOfDecember28 = int.parse(DateFormat("D").format(december28));
    return ((dayOfDecember28 - december28.weekday + 10) / 7).floor();
  }

  DateTime getFirstDayOfWeek(int weekNumber, int year) {
    var daysInYear = (weekNumber - 1) * 7;
    var firstDayOfFirstWeek = getFirstDayOfFirstWeek(year);
    return firstDayOfFirstWeek.add(new Duration(days: daysInYear));
  }

  DateTime getFirstDayOfFirstWeek(int year) {
    var firstDay = new DateTime(year, 1, 1);
    final numberOfWeeksInLastYear = getNumberOfWeeks(year - 1);
    if (numberOfWeeksInLastYear == 53) {
      final december28 = new DateTime(year - 1, 12, 28);
      // As devember 28th is always in the last wekk of the last year, a day one week later must be in the first week of the next year:
      firstDay = december28.add(new Duration(days: 7));
    }

    // If this day is not a monday, the first day of the week must be in the last year:
    if (firstDay.weekday != 1) {
      firstDay = firstDay.subtract(new Duration(days: firstDay.weekday - 1));
    }
    return firstDay;
  }

  DateTime getLastDayOfWeek(int weekNumber, int year) {
    var firstDayOfWeek = getFirstDayOfWeek(weekNumber, year);
    return firstDayOfWeek.add(new Duration(days: 6));
  }
}

File: server/src/pkg/tools/date_tools.go
package tools

import (
	"time"
)

func GetWeekNumber(date time.Time) int {
	dayOfYear := date.YearDay()
	weekNumber := (dayOfYear - int(date.Weekday()) + 10) / 7
	if weekNumber < 1 {
		weekNumber = GetNumberOfWeeks(date.Year() - 1)
	} else if weekNumber > GetNumberOfWeeks(date.Year()) {
		weekNumber = 1
	}
	return weekNumber
}

func GetNumberOfWeeks(year int) int {
	december28 := time.Date(year, time.December, 28, 0, 0, 0, 0, time.UTC)
	dayOfDecember28 := december28.YearDay()
	return (dayOfDecember28 - int(december28.Weekday()) + 10) / 7
}

func GetFirstDayOfWeek(weekNumber int, year int) time.Time {
	daysInYear := (weekNumber - 1) * 7
	firstDayOfFirstWeek := GetFirstDayOfFirstWeek(year)
	return firstDayOfFirstWeek.AddDate(0, 0, daysInYear)
}

func GetFirstDayOfFirstWeek(year int) time.Time {
	december28 := time.Date(year-1, time.December, 28, 0, 0, 0, 0, time.UTC)
	// As december 28th is always in the last week of the last year, a day one week later must be in the first week of the next year:
	firstDay := december28.AddDate(0, 0, 7)
	// If this day is not a monday, the first day of the week must be in the last year:
	if firstDay.Weekday() != time.Monday {
		daysFromMonday := (int(firstDay.Weekday()) - int(time.Monday) + 7) % 7
		firstDay = firstDay.AddDate(0, 0, -daysFromMonday)
	}
	return firstDay
}

func GetLastDayOfWeek(weekNumber int, year int) time.Time {
	firstDayOfWeek := GetFirstDayOfWeek(weekNumber, year)
	return firstDayOfWeek.AddDate(0, 0, 6)
}

func OnlyDate(dateTime time.Time) time.Time {
	return time.Date(dateTime.Year(), dateTime.Month(), dateTime.Day(), 0, 0, 0, 0, time.UTC)
}
